### ISSUE TITLE
Make HDF5 Outputs readable by yt

### ISSUE THREAD

**Issue by forrestglines** [parthenon-hpc-lab/parthenon#474]
For our astrophysics work with Parthenon, we'll need to be able to load Parthenon outputs with `yt` `ttps://github.com/yt-project/yt. Most of that work will be in developing a `yt` frontend for Parthenon, using the Athena++ frontend as a starting point, but it will likely require adding a few things to the Parthenon HDF5 data output. I believe we'll need to add:
- A 1D dataset with the AMR level of every meshblock
- A 2D dataset with the logical location of every meshblock within their AMR level
- Attributes for the root grid size on the 0th AMR level, how many mesh blocks are on the root grid, and the domain of the root grid
- Ability for downstream codes to embed attributes at the top level or within datasets. For AthenaPK, these would be a string for equation of state and a float for adiabatic index. Code unit information would also be useful.

Would these additions conflict with simultaneous work on performance?

**Comment by JoshuaSBrown** [comment]
I don't think so, I'm in the process of looking at the performance of the HDF5 io. Moving forward it may make sense to move to parallel MPI io instead of using the hdf5, but that has yet to be decided. If these attributes are going to be printed with every restart file it may make sense to make them optional to prevent unnecessary io overhead, others probably have more feedback.

**Comment by Yurlungur** [comment]
I think this would be great to have.

**Comment by forrestglines** [comment]
> I don't think so, I'm in the process of looking at the performance of the HDF5 io. Moving forward it may make sense to move to parallel MPI io instead of using the hdf5, but that has yet to be decided. 

If I need to reimplement the `yt` frontend for a better data format in the future, I'm fine with doing that. 

> If these attributes are going to be printed with every restart file it may make sense to make them optional to prevent unnecessary io overhead, others probably have more feedback.

The total additional data will be very minimal, a couple bytes per meshblock. Hopefully that won't add a performance cost, but I'll watch out for it and potentially make the additional data for yt optional but always written for AthenaPK.

**Comment by Yurlungur** [comment]
One thing to keep in mind, which we can worry about later, once we have a yt reader, is coordinate systems. I recall yt has pretty rich machinery for handling this, but we'd have to hook into it when the time comes.

**Comment by forrestglines** [comment]
I have a working prototype for `yt` outputs in #478  and a `yt` frontend at https://github.com/forrestglines/yt/tree/athenapk-frontend. Parts of my prototype need to be rolled into hooks for downstream codes and moved to AthenaPK.

I had to add a couple attributes and datasets. I'd like your opinion on whether we should write these by default, have a flag to toggle them on and off, or only include these downstream in AthenaPK (so that `yt` could only ready AthenaPK outputs).

New attributes:

- `RootGridDomain` Domain of the root grid (the coarsest AMR level)
- `RootGridSize` Number of cells on the root grid
- `BoundaryConditions` List of boundary conditions used, currently only used to determine periodicity
- `DatasetNames` Names of the datasets written, which are divided into variables. For Athena++, these were `cons`,`b`, etc.
- `NumVariables` Number of variables in each dataset
- `VariablesNames` Names of variables from all datasets. For Athena++, this would look like `dens`,`Etot`,`mom1`,`mom2`,`mom3`,`bcc1`,`bcc2`,`bcc3`

New datasets:

-`VolumeLocations` Locations of cell centers as opposed to face centers. It's not strictly necessary for uniform Cartesian coordinates, but it might be necessary for more complex geometries.
-`Levels` AMR level of each meshblock
-`LogicalLocations` Indexed position of each meshblock within their AMR level

We might also want to make hooks for downstream codes to add their own attributes. Although Athena++ didn't do this, I would like to add the adiabatic index, equation of state, and code units to the HDF5 output so that the output is self-contained.

**Comment by forrestglines** [comment]
Also, first output from `yt` with static mesh refinement. The lines show the meshblocks

![image](https://user-images.githubusercontent.com/5799640/112504559-493d5600-8d62-11eb-98fc-6f4013fb8c47.png)

**Comment by Yurlungur** [comment]
Nice work @forrestglines ! This looks great! I like all the new attributes/datasets. I also like the idea of hooks for downstream codes to add their own information.

**Comment by Yurlungur** [comment]
One thing to be careful of is the yt reader should be robust to missing inputs from downstream codes. i.e., if AthenaPK and Riot both write to yt, the adiabatic index will be almost certainly missing from riot.

**Comment by forrestglines** [comment]
> One thing to be careful of is the yt reader should be robust to missing inputs from downstream codes. i.e., if AthenaPK and Riot both write to yt, the adiabatic index will be almost certainly missing from riot.

I agree, especially about the adiabatic index. After all, adiabatic index only makes sense for some equations of state.

One piece concerning compatibility though is that the `yt` reader will expect very specific names for density, momenta, etc. These are used internally by `yt` to compute derived fields, so it will break if they can't find them. Right now my `yt` frontend expects the same names that Athena++ used (`dens`,`Etot`, etc.) but we could also use more descriptive names instead (i.e. `Density`,`TotalEnergy`, etc. ) which would make more sense in a downstream code like Riot. In any case, Riot at some point would need to hook into the `yt` output to define how variables are arranged in the dataset.

**Comment by Yurlungur** [comment]
Hm... Could we perhaps define a name map in the metdata specified by downstream code, which tells yt which variable is which?

**Comment by Yurlungur** [comment]
@forrestglines what's the status of this? Is the parthenon frontend in `yt` proper or is it still in a branch?

**Comment by pgrete** [comment]
Still in a branch though we're getting close (need final testing for other coordinate system than Cartesian and adding/pushing/integrating the test cases).
